File: typescript/src/binary.ts
// Wire-format binary readers — keep in sync with
// `app/schemas/points_binary.py` and `app/schemas/depth_map_binary.py`.

const MAGIC_POINTS = [0x53, 0x46, 0x4d, 0x50, 0x33, 0x44, 0x00, 0x00]; // "SFMP3D\0\0"
const MAGIC_DEPTH = [0x53, 0x46, 0x4d, 0x44, 0x50, 0x54, 0x48, 0x00]; // "SFMDPTH\0"
const MAGIC_NORMAL = [0x53, 0x46, 0x4d, 0x4e, 0x52, 0x4d, 0x00, 0x00]; // "SFMNRM\0\0"

const POINTS_HEADER_SIZE = 44;
const POINTS_RECORD_SIZE = 26;
const MAP_HEADER_SIZE = 32;

function _eqMagic(view: DataView, expected: number[]): boolean {
  for (let i = 0; i < expected.length; i++) {
    if (view.getUint8(i) !== expected[i]) return false;
  }
  return true;
}

// ---- application/x-sfm-points-v1 ----------------------------------------

export interface Point3DRecord {
  point3d_id: bigint;
  xyz: [number, number, number];
  rgb: [number, number, number];
  track_len: number;
}

export interface PointsBinary {
  count: number;
  bbox_min: [number, number, number];
  bbox_max: [number, number, number];
  records: Point3DRecord[];
}

export function parsePointsBinary(buffer: ArrayBuffer): PointsBinary {
  if (buffer.byteLength < POINTS_HEADER_SIZE) {
    throw new Error(`points-binary: buffer too small for header`);
  }
  const view = new DataView(buffer);
  if (!_eqMagic(view, MAGIC_POINTS)) {
    throw new Error("points-binary: bad magic");
  }
  const version = view.getUint32(8, true);
  if (version !== 1) {
    throw new Error(`points-binary: unknown version ${version}`);
  }
  const count = Number(view.getBigUint64(12, true));
  const bboxMin: [number, number, number] = [
    view.getFloat32(20, true),
    view.getFloat32(24, true),
    view.getFloat32(28, true),
  ];
  const bboxMax: [number, number, number] = [
    view.getFloat32(32, true),
    view.getFloat32(36, true),
    view.getFloat32(40, true),
  ];
  const expected = POINTS_HEADER_SIZE + count * POINTS_RECORD_SIZE;
  if (buffer.byteLength < expected) {
    throw new Error(
      `points-binary: body short — got ${buffer.byteLength}, expected ${expected}`,
    );
  }
  const records: Point3DRecord[] = new Array(count);
  for (let i = 0; i < count; i++) {
    const off = POINTS_HEADER_SIZE + i * POINTS_RECORD_SIZE;
    records[i] = {
      xyz: [
        view.getFloat32(off, true),
        view.getFloat32(off + 4, true),
        view.getFloat32(off + 8, true),
      ],
      rgb: [
        view.getUint8(off + 12),
        view.getUint8(off + 13),
        view.getUint8(off + 14),
      ],
      track_len: view.getUint16(off + 16, true),
      point3d_id: view.getBigUint64(off + 18, true),
    };
  }
  return { count, bbox_min: bboxMin, bbox_max: bboxMax, records };
}

// ---- application/x-sfm-depth-v1 -----------------------------------------

export interface DepthMap {
  width: number;
  height: number;
  depth_min: number;
  depth_max: number;
  /** Row-major float32 array of length `width * height`. */
  pixels: Float32Array;
}

export function parseDepthMap(buffer: ArrayBuffer): DepthMap {
  if (buffer.byteLength < MAP_HEADER_SIZE) {
    throw new Error("depth-binary: buffer too small for header");
  }
  const view = new DataView(buffer);
  if (!_eqMagic(view, MAGIC_DEPTH)) {
    throw new Error("depth-binary: bad magic");
  }
  const version = view.getUint32(8, true);
  if (version !== 1) throw new Error(`depth-binary: unknown version ${version}`);
  const width = view.getUint32(12, true);
  const height = view.getUint32(16, true);
  const depthMin = view.getFloat32(20, true);
  const depthMax = view.getFloat32(24, true);
  const pixelCount = width * height;
  const expected = MAP_HEADER_SIZE + pixelCount * 4;
  if (buffer.byteLength < expected) {
    throw new Error(
      `depth-binary: body short — got ${buffer.byteLength}, expected ${expected}`,
    );
  }
  const pixels = new Float32Array(buffer, MAP_HEADER_SIZE, pixelCount);
  return { width, height, depth_min: depthMin, depth_max: depthMax, pixels };
}

// ---- application/x-sfm-normal-v1 ----------------------------------------

export interface NormalMap {
  width: number;
  height: number;
  /** Row-major channels-last float32 of length `width * height * 3`. */
  pixels: Float32Array;
}

export function parseNormalMap(buffer: ArrayBuffer): NormalMap {
  if (buffer.byteLength < MAP_HEADER_SIZE) {
    throw new Error("normal-binary: buffer too small for header");
  }
  const view = new DataView(buffer);
  if (!_eqMagic(view, MAGIC_NORMAL)) {
    throw new Error("normal-binary: bad magic");
  }
  const version = view.getUint32(8, true);
  if (version !== 1) throw new Error(`normal-binary: unknown version ${version}`);
  const width = view.getUint32(12, true);
  const height = view.getUint32(16, true);
  const expected = MAP_HEADER_SIZE + width * height * 3 * 4;
  if (buffer.byteLength < expected) {
    throw new Error(
      `normal-binary: body short — got ${buffer.byteLength}, expected ${expected}`,
    );
  }
  const pixels = new Float32Array(buffer, MAP_HEADER_SIZE, width * height * 3);
  return { width, height, pixels };
}

File: python/sfmapi_client/binary.py
"""Wire-format binary decoders for the sfmapi Python SDK.

Mirrors ``app/schemas/points_binary.py`` (server-side encoder) and
``app/schemas/depth_map_binary.py``. Pure-stdlib (struct + memoryview);
no NumPy required.
"""

from __future__ import annotations

import struct
from dataclasses import dataclass

POINTS_MAGIC = b"SFMP3D\x00\x00"
DEPTH_MAGIC = b"SFMDPTH\x00"
NORMAL_MAGIC = b"SFMNRM\x00\x00"

POINTS_HEADER_SIZE = 44
POINTS_RECORD_SIZE = 26
MAP_HEADER_SIZE = 32

POINTS_MEDIA_TYPE = "application/x-sfm-points-v1"
DEPTH_MEDIA_TYPE = "application/x-sfm-depth-v1"
NORMAL_MEDIA_TYPE = "application/x-sfm-normal-v1"


class WireFormatError(ValueError):
    """Raised when a binary blob fails magic / version / length checks."""


@dataclass(frozen=True)
class Point3DRecord:
    point3d_id: int
    xyz: tuple[float, float, float]
    rgb: tuple[int, int, int]
    track_len: int


@dataclass
class PointsBinary:
    count: int
    bbox_min: tuple[float, float, float]
    bbox_max: tuple[float, float, float]
    records: list[Point3DRecord]


def parse_points_binary(data: bytes) -> PointsBinary:
    """Decode an ``application/x-sfm-points-v1`` blob."""
    if len(data) < POINTS_HEADER_SIZE:
        raise WireFormatError("points-binary: buffer too small for header")
    if data[:8] != POINTS_MAGIC:
        raise WireFormatError("points-binary: bad magic")
    version = struct.unpack_from("<I", data, 8)[0]
    if version != 1:
        raise WireFormatError(f"points-binary: unknown version {version}")
    count = struct.unpack_from("<Q", data, 12)[0]
    bbox_min = struct.unpack_from("<fff", data, 20)
    bbox_max = struct.unpack_from("<fff", data, 32)
    expected = POINTS_HEADER_SIZE + count * POINTS_RECORD_SIZE
    if len(data) < expected:
        raise WireFormatError(f"points-binary: body short — got {len(data)}, expected {expected}")
    records: list[Point3DRecord] = []
    for i in range(count):
        off = POINTS_HEADER_SIZE + i * POINTS_RECORD_SIZE
        x, y, z, r, g, b, _pad, tl, pid = struct.unpack_from("<fffBBBBHQ", data, off)
        records.append(
            Point3DRecord(
                point3d_id=pid,
                xyz=(x, y, z),
                rgb=(r, g, b),
                track_len=tl,
            )
        )
    return PointsBinary(count=count, bbox_min=bbox_min, bbox_max=bbox_max, records=records)


@dataclass
class DepthMap:
    width: int
    height: int
    depth_min: float
    depth_max: float
    pixels: bytes  # raw float32 bytes (caller decodes with struct or array.array)


def parse_depth_map(data: bytes) -> DepthMap:
    """Decode an ``application/x-sfm-depth-v1`` blob.

    ``pixels`` is the raw float32 little-endian byte buffer of length
    ``width * height * 4``. Use ``struct`` or ``array.array("f")`` to
    materialize it (or ``np.frombuffer(pixels, dtype="<f4")`` if NumPy
    is available)."""
    if len(data) < MAP_HEADER_SIZE:
        raise WireFormatError("depth-binary: buffer too small for header")
    if data[:8] != DEPTH_MAGIC:
        raise WireFormatError("depth-binary: bad magic")
    version = struct.unpack_from("<I", data, 8)[0]
    if version != 1:
        raise WireFormatError(f"depth-binary: unknown version {version}")
    w, h = struct.unpack_from("<II", data, 12)
    dmin, dmax = struct.unpack_from("<ff", data, 20)
    expected = MAP_HEADER_SIZE + w * h * 4
    if len(data) < expected:
        raise WireFormatError(f"depth-binary: body short — got {len(data)}, expected {expected}")
    pixels = bytes(data[MAP_HEADER_SIZE : MAP_HEADER_SIZE + w * h * 4])
    return DepthMap(width=w, height=h, depth_min=dmin, depth_max=dmax, pixels=pixels)


@dataclass
class NormalMap:
    width: int
    height: int
    pixels: bytes  # raw float32 bytes, length = width*height*3*4


def parse_normal_map(data: bytes) -> NormalMap:
    """Decode an ``application/x-sfm-normal-v1`` blob."""
    if len(data) < MAP_HEADER_SIZE:
        raise WireFormatError("normal-binary: buffer too small for header")
    if data[:8] != NORMAL_MAGIC:
        raise WireFormatError("normal-binary: bad magic")
    version = struct.unpack_from("<I", data, 8)[0]
    if version != 1:
        raise WireFormatError(f"normal-binary: unknown version {version}")
    w, h = struct.unpack_from("<II", data, 12)
    expected = MAP_HEADER_SIZE + w * h * 3 * 4
    if len(data) < expected:
        raise WireFormatError(f"normal-binary: body short — got {len(data)}, expected {expected}")
    pixels = bytes(data[MAP_HEADER_SIZE : MAP_HEADER_SIZE + w * h * 3 * 4])
    return NormalMap(width=w, height=h, pixels=pixels)


__all__ = [
    "DEPTH_MAGIC",
    "DEPTH_MEDIA_TYPE",
    "MAP_HEADER_SIZE",
    "NORMAL_MAGIC",
    "NORMAL_MEDIA_TYPE",
    "POINTS_HEADER_SIZE",
    "POINTS_MAGIC",
    "POINTS_MEDIA_TYPE",
    "POINTS_RECORD_SIZE",
    "DepthMap",
    "NormalMap",
    "Point3DRecord",
    "PointsBinary",
    "WireFormatError",
    "parse_depth_map",
    "parse_normal_map",
    "parse_points_binary",
]
